File: backend/pkg/core/worker/discovery.go
package worker

import (
	"context"
	"fmt"
	"log/slog"
	"net/url"
	"strings"

	"backend/pkg/core/datasource"
	"backend/pkg/database"
	"backend/pkg/database/repository"
	"backend/pkg/integrations/exa"
	"backend/pkg/types"
)

// DiscoveryWorker discovers similar websites using Exa
type DiscoveryWorker struct {
	logger           *slog.Logger
	exaClient        *exa.Client
	maxDiscoveries   int
	maxSourcesPerRun int
}

// NewDiscoveryWorker creates a new DiscoveryWorker instance
func NewDiscoveryWorker(logger *slog.Logger, exaAPIKey string) *DiscoveryWorker {
	if logger == nil {
		logger = slog.Default()
	}

	var client *exa.Client
	if exaAPIKey != "" {
		client = exa.NewClient(exaAPIKey)
	}

	return &DiscoveryWorker{
		logger:           logger,
		exaClient:        client,
		maxDiscoveries:   5, // Max new sources to discover per existing source
		maxSourcesPerRun: 5, // Max sources to process per run
	}
}

// Name returns the worker name
func (w *DiscoveryWorker) Name() string {
	return "discovery"
}

// Run executes the discovery worker
func (w *DiscoveryWorker) Run(ctx context.Context) error {
	w.logger.Info("starting discovery worker run")
	statusService := GetStatusService()

	if w.exaClient == nil {
		w.logger.Warn("Exa client not configured, skipping discovery")
		statusService.UpdateStatus(w.Name(), StateRunning, 100, "Exa not configured, skipping")
		return nil
	}

	// Get active data sources (not discovered ones, as they're already derived)
	statusService.UpdateStatus(w.Name(), StateRunning, 0, "Fetching data sources...")
	repo := repository.NewDataSourceRepository(database.DB())
	sources, _, err := repo.List(ctx, 0, 100)
	if err != nil {
		return fmt.Errorf("failed to get data sources: %w", err)
	}

	// Filter to only user-added sources
	var activeManualSources []types.DataSource
	for _, s := range sources {
		if s.IsEnabled && !s.IsDiscovered {
			activeManualSources = append(activeManualSources, s)
		}
	}

	if len(activeManualSources) == 0 {
		w.logger.Info("no active manual data sources found")
		statusService.UpdateStatus(w.Name(), StateRunning, 100, "No manual sources to discover from")
		return nil
	}

	// Limit sources to process
	if len(activeManualSources) > w.maxSourcesPerRun {
		activeManualSources = activeManualSources[:w.maxSourcesPerRun]
	}

	w.logger.Info("discovering similar websites", "source_count", len(activeManualSources))
	statusService.SetProgress(w.Name(), 0, len(activeManualSources), fmt.Sprintf("Processing %d sources", len(activeManualSources)))

	totalDiscovered := 0
	for i, source := range activeManualSources {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
			statusService.SetProgress(w.Name(), i, len(activeManualSources), fmt.Sprintf("Finding similar sites for: %s", source.Name))
			
			discovered, err := w.discoverSimilarSites(ctx, &source)
			if err != nil {
				w.logger.Error("failed to discover similar sites", "source_id", source.ID, "error", err)
				continue
			}
			totalDiscovered += discovered
		}
	}

	w.logger.Info("discovery worker completed", "total_discovered", totalDiscovered)
	statusService.SetProgress(w.Name(), len(activeManualSources), len(activeManualSources), fmt.Sprintf("Discovered %d new sites", totalDiscovered))
	return nil
}

// discoverSimilarSites discovers similar websites for a given source
func (w *DiscoveryWorker) discoverSimilarSites(ctx context.Context, source *types.DataSource) (int, error) {
	w.logger.Info("discovering similar sites", "source_id", source.ID, "source_url", source.URL)

	// Use Exa to find similar sites
	results, err := w.exaClient.FindSimilar(ctx, source.URL, &exa.FindSimilarOptions{
		NumResults:        w.maxDiscoveries + 5, // Get extra to account for filtering
		ExcludeSourceDomain: true,
	})
	if err != nil {
		return 0, fmt.Errorf("Exa findSimilar failed: %w", err)
	}

	if len(results.Results) == 0 {
		w.logger.Info("no similar sites found", "source_id", source.ID)
		return 0, nil
	}

	discoveredCount := 0
	for _, result := range results.Results {
		if discoveredCount >= w.maxDiscoveries {
			break
		}

		// Skip if URL looks invalid
		if result.URL == "" || !strings.HasPrefix(result.URL, "http") {
			continue
		}

		// Normalize URL (remove trailing slashes, etc.)
		normalizedURL := normalizeURL(result.URL)
		if normalizedURL == "" {
			continue
		}

		// Skip if it's too similar to source (same domain root)
		if isSameDomainRoot(source.URL, normalizedURL) {
			continue
		}

		// Create discovered data source
		name := result.Title
		if name == "" {
			name = extractDomainName(normalizedURL)
		}

		// Pass both orgID and userID from the source
		_, err := datasource.CreateDiscoveredSource(ctx, source.OrganizationID, source.UserID, source.ID, name, normalizedURL)
		if err != nil {
			// Skip if already exists
			if strings.Contains(err.Error(), "already exists") {
				continue
			}
			w.logger.Warn("failed to create discovered source", "url", normalizedURL, "error", err)
			continue
		}

		discoveredCount++
		w.logger.Info("discovered similar site", "url", normalizedURL, "name", name, "from_source", source.ID)
	}

	return discoveredCount, nil
}

// Helper functions

func normalizeURL(rawURL string) string {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return ""
	}

	// Remove trailing slash
	parsed.Path = strings.TrimSuffix(parsed.Path, "/")

	// Remove fragment
	parsed.Fragment = ""

	// Remove common tracking parameters
	query := parsed.Query()
	for key := range query {
		lowerKey := strings.ToLower(key)
		if strings.HasPrefix(lowerKey, "utm_") ||
			lowerKey == "ref" ||
			lowerKey == "source" ||
			lowerKey == "campaign" {
			query.Del(key)
		}
	}
	parsed.RawQuery = query.Encode()

	return parsed.String()
}

func isSameDomainRoot(url1, url2 string) bool {
	root1 := extractDomainRoot(url1)
	root2 := extractDomainRoot(url2)
	return root1 != "" && root2 != "" && root1 == root2
}

func extractDomainRoot(rawURL string) string {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return ""
	}

	host := strings.ToLower(parsed.Host)

	// Remove www prefix
	host = strings.TrimPrefix(host, "www.")

	// Extract domain root (e.g., "example.com" from "blog.example.com")
	parts := strings.Split(host, ".")
	if len(parts) >= 2 {
		return parts[len(parts)-2] + "." + parts[len(parts)-1]
	}

	return host
}

func extractDomainName(rawURL string) string {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return rawURL
	}

	host := strings.TrimPrefix(parsed.Host, "www.")

	// Try to create a readable name from domain
	parts := strings.Split(host, ".")
	if len(parts) > 0 {
		// Capitalize first part
		name := parts[0]
		if len(name) > 0 {
			return strings.ToUpper(string(name[0])) + name[1:]
		}
	}

	return host
}

File: backend/src/core/worker/discovery.rs
use std::sync::Arc;

use async_trait::async_trait;
use reqwest::Url;
use serde_json::json;

use crate::{
    core::datasource::{
        DataSourceResponse, DataSourceService, RecommendationSearchPort, SimilarOptions,
    },
    error::AppError,
};

use super::{StatusService, Worker, WorkerContext, WorkerFailure, WorkerResult, WorkerState};

const DISCOVERY_WORKER_NAME: &str = "discovery";
const DEFAULT_MAX_DISCOVERIES: i32 = 5;
const DEFAULT_MAX_SOURCES: usize = 5;

pub struct DiscoveryWorker {
    status: Arc<StatusService>,
    data_sources: Arc<DataSourceService>,
    search: Option<Arc<dyn RecommendationSearchPort>>,
    max_discoveries: i32,
    max_sources: usize,
}

impl DiscoveryWorker {
    pub fn new(
        status: Arc<StatusService>,
        data_sources: Arc<DataSourceService>,
        search: Option<Arc<dyn RecommendationSearchPort>>,
    ) -> Self {
        Self {
            status,
            data_sources,
            search,
            max_discoveries: DEFAULT_MAX_DISCOVERIES,
            max_sources: DEFAULT_MAX_SOURCES,
        }
    }

    async fn discover_similar(
        &self,
        search: &dyn RecommendationSearchPort,
        source: &DataSourceResponse,
    ) -> Result<i32, WorkerFailure> {
        let response = search
            .find_similar(
                &source.url,
                SimilarOptions {
                    num_results: self.max_discoveries + 5,
                    exclude_source_domain: true,
                    ..SimilarOptions::default()
                },
            )
            .await
            .map_err(|error| WorkerFailure::new(format!("Exa findSimilar failed: {error}")))?;
        let mut discovered = 0_i32;
        for result in response.results {
            if discovered >= self.max_discoveries {
                break;
            }
            let Some(url) = normalize_url(&result.url) else {
                continue;
            };
            if is_same_domain_root(&source.url, &url) {
                continue;
            }
            let name = if result.title.is_empty() {
                extract_domain_name(&url)
            } else {
                result.title
            };
            match self
                .data_sources
                .create_discovered_source(
                    source.organization_id,
                    source.user_id,
                    source.id,
                    name,
                    url,
                )
                .await
            {
                Ok(_) => discovered += 1,
                Err(AppError::Conflict(_)) => {}
                Err(_) => {}
            }
        }
        Ok(discovered)
    }
}

#[async_trait]
impl Worker for DiscoveryWorker {
    fn name(&self) -> &str {
        DISCOVERY_WORKER_NAME
    }

    async fn run(&self, context: WorkerContext) -> Result<WorkerResult, WorkerFailure> {
        let Some(search) = self.search.as_ref().filter(|search| search.is_configured()) else {
            self.status.update_status(
                self.name(),
                WorkerState::Running,
                100,
                "Exa not configured, skipping",
            );
            return Ok(WorkerResult::warning(
                "Exa not configured, discovery skipped",
                vec!["Exa client not configured".to_owned()],
            ));
        };
        self.status.update_status(
            self.name(),
            WorkerState::Running,
            0,
            "Fetching data sources...",
        );
        let (sources, _) = tokio::select! {
            biased;
            () = context.cancelled() => return Err(WorkerFailure::new("operation cancelled")),
            result = self.data_sources.list_all(1, 100) => {
                result.map_err(|error| WorkerFailure::new(format!("failed to get data sources: {error}")))?
            }
        };
        let sources = sources
            .into_iter()
            .filter(|source| source.is_enabled && !source.is_discovered)
            .take(self.max_sources)
            .collect::<Vec<_>>();
        if sources.is_empty() {
            self.status.update_status(
                self.name(),
                WorkerState::Running,
                100,
                "No manual sources to discover from",
            );
            return Ok(WorkerResult::warning(
                "No manual sources to discover from",
                vec!["No enabled manual sources were available".to_owned()],
            ));
        }

        let total = i32::try_from(sources.len())
            .map_err(|_| WorkerFailure::new("too many discovery sources"))?;
        self.status
            .set_progress(self.name(), 0, total, format!("Processing {total} sources"));
        let mut total_discovered = 0_i32;
        for (index, source) in sources.into_iter().enumerate() {
            if context.cancellation().is_cancelled() {
                return Err(WorkerFailure::new("operation cancelled"));
            }
            let done = i32::try_from(index)
                .map_err(|_| WorkerFailure::new("too many discovery sources"))?;
            self.status.set_progress(
                self.name(),
                done,
                total,
                format!("Finding similar sites for: {}", source.name),
            );
            if let Ok(discovered) = self.discover_similar(search.as_ref(), &source).await {
                total_discovered = total_discovered.saturating_add(discovered);
            }
        }
        self.status.set_progress(
            self.name(),
            total,
            total,
            format!("Discovered {total_discovered} new sites"),
        );
        let mut result = if total_discovered == 0 {
            WorkerResult::warning(
                "Discovery completed without new sites",
                vec!["No new adjacent sites were discovered".to_owned()],
            )
        } else {
            WorkerResult::completed(format!("Discovered {total_discovered} new sites"))
        };
        result
            .metrics
            .insert("seed_sources".to_owned(), json!(total));
        result
            .metrics
            .insert("discovered_sources".to_owned(), json!(total_discovered));
        if total_discovered > 0 {
            result
                .output_summary
                .insert("discovered_sources".to_owned(), json!(total_discovered));
        }
        Ok(result)
    }
}

pub fn normalize_url(raw_url: &str) -> Option<String> {
    let mut url = Url::parse(raw_url).ok()?;
    if !matches!(url.scheme(), "http" | "https") {
        return None;
    }
    let trimmed_path = url.path().trim_end_matches('/').to_owned();
    url.set_path(&trimmed_path);
    url.set_fragment(None);
    let query = url
        .query_pairs()
        .filter(|(key, _)| {
            let key = key.to_ascii_lowercase();
            !key.starts_with("utm_") && !matches!(key.as_str(), "ref" | "source" | "campaign")
        })
        .map(|(key, value)| (key.into_owned(), value.into_owned()))
        .collect::<Vec<_>>();
    url.query_pairs_mut().clear().extend_pairs(query);
    if url.query() == Some("") {
        url.set_query(None);
    }
    Some(url.to_string())
}

fn is_same_domain_root(left: &str, right: &str) -> bool {
    match (domain_root(left), domain_root(right)) {
        (Some(left), Some(right)) => left == right,
        _ => false,
    }
}

fn domain_root(value: &str) -> Option<String> {
    let host = Url::parse(value).ok()?.host_str()?.to_ascii_lowercase();
    let host = host.strip_prefix("www.").unwrap_or(&host);
    let parts = host.split('.').collect::<Vec<_>>();
    if parts.len() >= 2 {
        Some(format!(
            "{}.{}",
            parts[parts.len() - 2],
            parts[parts.len() - 1]
        ))
    } else {
        Some(host.to_owned())
    }
}

fn extract_domain_name(value: &str) -> String {
    let Some(host) = Url::parse(value)
        .ok()
        .and_then(|url| url.host_str().map(str::to_owned))
    else {
        return value.to_owned();
    };
    let host = host.strip_prefix("www.").unwrap_or(&host);
    let name = host.split('.').next().unwrap_or(host);
    let mut characters = name.chars();
    match characters.next() {
        Some(first) => first.to_uppercase().chain(characters).collect(),
        None => host.to_owned(),
    }
}
